test/testutil/storage/std_list_adapter.hpp

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_STD_LIST_ADAPTER_HPP
+#define KAGOME_STD_LIST_ADAPTER_HPP
+
+#include <list>
+
+#include "storage/face/generic_iterator.hpp"
+#include "storage/face/generic_list.hpp"
+
+namespace test {
+
+  using kagome::face::ForwardIterator;
+  using kagome::face::GenericIterator;
+  using kagome::face::GenericList;
+
+  /**
+   * An adapter of std::list iterator for face::GenericIterator interface
+   * @tparam T type of an element stored in the list
+   */
+  template <typename T>
+  class StdListIterator : public GenericIterator<GenericList<T>> {
+    template <typename>
+    friend class StdListAdapter;
+
+   public:
+    using value_type = T;
+
+    explicit StdListIterator(typename std::list<T>::iterator l)
+        : it_{std::move(l)} {}
+    ~StdListIterator() override = default;
+
+    std::unique_ptr<GenericIterator<GenericList<T>>> create_copy()
+        const override {
+      return std::make_unique<StdListIterator<T>>(it_);
+    }
+
+    value_type *get() override {
+      return &*it_;
+    }
+
+    value_type const *get() const override {
+      return &*it_;
+    }
+
+    value_type &operator*() override {
+      return *it_;
+    }
+
+    value_type const &operator*() const override {
+      return *it_;
+    }
+
+    kagome::face::GenericIterator<GenericList<T>> &operator++() override {
+      it_++;
+      return *this;
+    }
+
+   private:
+    typename std::list<T>::iterator it_;
+  };
+
+  /**
+   * An adapter of std::list for face::GenericList interface
+   * @tparam T type of an element stored in the list
+   */
+  template <typename T>
+  class StdListAdapter : public GenericList<T> {
```

Comment:
what is the point of this, generic list and generic iterator here?

Reply:
To pass a container to TransactionPool through an interface, to easily replace, for example, an std::list with a thread-safe container, or mock it in tests.

Reply:
if you want to generify a container, wouldn't templates work better here?

Reply:
Don't want to make TransactionPool a template.

Reply:
Why not?

Reply:
Because that would introduce some unnecessary complexity, for example it would complicate separation on header and source, and also any use of TransactionPool type would require specifying this template parameter, which I don't like as it is a part of the inner implementation of the pool and therefore it's not meant to be brought up everywhere the transaction pool is used.

Reply:
what template does here - it just adds a possibility to change container. 
you then can create an alias at some place, which uses some default container.

like
```
template <typename T>
class GenericPool {
 T container;
};

// and then use Pool everywhere
using Pool = GenericPool<std::list>;
```

Reply:
or even better - you can do explicit instantiation of `GenericPool<std::list>` in some cpp, this will decrease compile times